Add a document with guidelines for picking SLIs (#8)
operate-first/SRE: picking_good_slis.md

# Picking Good SLIs

## SRE's Philosophy on Monitoring

A core goal of Site Reliability Engineering (SRE) is to ensure the availability of a service. "Availability" here means much more than simply "is my application running." Effectively monitoring a modern cloud application requires insight into the application's response time, error rate, traffic level, and more.

> Furthermore, the scope of monitoring should not be limited solely for use in deciding when to page an operations team. Metrics should be closely tied to business objectives, informing decisions on what features to prioritize and when to focus on adding new features versus improving reliability.

Reviewer: ```suggestion
Furthermore, the scope of monitoring should not be limited solely for use in deciding when to page an operations team. Metrics should be closely tied to business objectives, informing decisions on what work to prioritize and when to focus on adding new features versus improving reliability.
```

Do you mean "work" here and not "features?"

Author: Fixed.

> To make this vision a reality, SRE teams rely on two key concepts:

Reviewer: Is saying "SRE teams" here unnecessarily narrow in scope? Is it reasonable to expand the scope to "service teams" (or similar) with the goal of looping in the dev team(s) reasponsible for the service?

My thinking is that in the forthcoming SLO RACI chart, the personas responsible for executing the SLO roadmap are both Dev and SRE ICs. The execution of the roadmap may include instrumenting the service in question, and I would think that would best fall to the development team of that service.

Reviewer 2: yeah, s/SRE/service

Author: Updated.

>   * Service Level Indicators (SLIs) - A quantifiable metric that measures a specific aspect of a service's level of service.

Reviewer 2: I am noticing that we define SLI/SLO several times in various docs.  I think once the dust settles we should go through and have either a single place where we define it or make sure all places where they're defined say the same exact thing. an #includes type of feature here would be nice ...

  * Service Level Objectives (SLOs) - A measurable, specific goal for a given service's level of service over time, based on one or more SLIs.

> In starting on this journey, we recommend first identifying SLOs for your service. In doing so, consider the perspective of the end user and what aspects of level of service are most important to them. The details that follow offer guidance for selecting the best SLIs to measure adherence to your SLOs.

Reviewer 2: link to lifecycle doc.

Author: Done.

## Selecting SLIs

In choosing SLIs, we recommend attempting to implement a measure for each of the four "Golden Signals".

### The Four Golden Signals of Monitoring

> Google defines the four golden signals well in their SRE Book: "The four golden signals of monitoring are latency, traffic, errors, and saturation. If you can only measure four metrics of your user-facing system, focus on these four."

Reviewer 2: link directly to their book here.

Author: Done.

#### Saturation

Saturation measures the utilization of your service relative to that point at which it would be considered "full". Effectively implementing SLIs for saturation requires knowledge, or at least a hypothesis, of the capacity for your service and the limiting factor. Examples of how to measure this might be:
  * CPU utilization percentage for services that are highly CPU-constrained
  * Memory utilization percentage for services that are highly Memory constrained
  * Storage capacity usage percentage
  * Network bandwidth capacity usage percentage
  * API request count as a percentage of maximum capacity

Indicators around saturation can be used as a safeguard in a well managed environment where limits are not expected to be reached. However, having unused/wasted resources is not ideal, so you may want to increase capacity on demand. Whether that means a manual increase or an automated scaling of instances, the important piece is having the right indicator of saturation for your service i.e. what is likely to be exhausted first as users use my service.

### Common Pitfalls to Avoid

> While not exhaustive, here are a few commonly-made mistakes that we recommend against:

Reviewer 3: We're missing the bullet points here

Author: @dastergon the intent was that the sub-headings are the bullet points.

# Picking Good SLIs

## SRE's Philosophy on Monitoring

> A core goal of SRE is to ensure the availability of a service. "Availability"

Reviewer: > SRE

We may want to define it here, like

```suggestion
A core goal of Site Reliability Engineering (SRE) is to ensure the availability of a service. "Availability"
```

or ...Site Reliability Engineers (SRE)...

Author: Updated in google doc.

here means much more than simply "is my application running." Effectively
monitoring a modern cloud application requires insight into the application's
response time, error rate, traffic level, and more.

Furthermore, the scope of monitoring should not be limited solely for use in
deciding when to page an operations team. Metrics should be closely tied to
business objectives, informing decisions on what features to prioritize and
when to focus on adding new features versus improving reliability.

To make this vision a reality, SRE teams rely on two key concepts:
  * Service Level Indicators (SLIs) - A quantifiable metric that measures a

>     specific aspect of a service's availability.

Reviewer 4: ```suggestion
    specific aspect of a service's level of service.
```

The definition from the [Google SRE book](https://sre.google/sre-book/service-level-objectives/) is 

> An SLI is a service level indicator—a carefully defined quantitative measure of some aspect of the level of service that is provided.

Avoiding the word availability is key here in my opinion which has been well explained in the first paragraph of this  document.

Author: Addressed in google doc.

  * Service Level Objectives (SLOs) - A measurable, specific goal for a

>     given service's availability over time, based on one or more SLIs.

Reviewer 4: ```suggestion
    given service's level of service over time, based on one or more SLIs.
```

Author: Addressed in google doc.

As a starting point on towards realizing this goal, this document aims to

> provide guidance on selecting the right SLIs for your service.

Reviewer: ```suggestion
This document aims to provide guidance on selecting the right SLIs for your
service.
```

The first part of the original sentence is a little cumbersome to read and as I was thinking of a suggestion for that part specifically, I thought to offer this suggestion instead. What do you think?

Author: Update in Google doc.

## Selecting SLIs

For SRE teams looking to start monitoring their services, start by identifying

> SLIs to measure. Focus on measuring the basics to start, can refine them over

Reviewer 4: I have added the point in a below comment with more details. I believe it is more useful to think through the SLOs that are most important for an end user of the service and then think through the SLIs.

As a separate effort, we can think through other SLIs which would be useful internally that would not have an associated SLO

Author: I copied this comment into the Google doc for further discussion.

> SLIs to measure. Focus on measuring the basics to start, can refine them over

Reviewer: This seems somewhat self-referential. The section is titled "Selecting SLIs" and then mentions that SRE teams should start by "identifying SLIs to measure." I agree with David's comment that a top-down approach may be better for teams new to SLOs and SLIs where if the business needs are the focus then the objective metrics can be identified to support them.

Separately, the second sentence seems to be lacking a subject. Who should focus on measuring the basics to start? Who can refine them (SLIs, presumably) over time? I think attaching a subject to the sentence will better help to convey your meaning.

Author: I'll copy your first comment here as a reply to David's note on this in the Google doc.

For the second point, after reading it again I don't think the point I was trying to convey really adds any value, so I'll just delete it.

time as you gain more experience operating the service.

A good starting point is to attempt to implement an SLI for each of the four
"Golden Signals".

> ## The Four Golden Signals of Monitoring

Reviewer: This section is a good summary of Google's SRE book. I'd like to see some advice on the mechanics of instrumentation to guide teams.

One of the things that comes to mind immediately would be a caution against "long inferences." A "long inference" can be explained by analogy: Measuring a laptop's CPU temperature by using the CD-ROM thermal sensor will not yield a CPU temperature measurement that has high fidelity.

This document is the perfect place to capture these kinds of pitfalls for the various kinds of golden signals.

Some other pitfalls that could be explored:

* Where to start and stop a stopwatch for latency measurements (when is the request "over"?)
* Long long are messages in a message queue before being processed vs message queue ingestion rate

Finally, I'd like to see a caution against hooking up metrics to alerts for the sake of doing _something_ at the start of an SRE team's journey. This goes to the "top-down thinking" comment I made elsewhere: if SLOs drive SLIs, then it is less likely that alerts are attached to things that "don't matter."

Google defines the four golden signals well in their [SRE Book][1]: "The four
golden signals of monitoring are latency, traffic, errors, and saturation. If
you can only measure four metrics of your user-facing system, focus on these
four."

We recommend implementing at least one SLI for each of these signals.

### Latency

Latency measures how long it takes your service to handle a request. Examples
of how to measure this might be:

>   * How long does it take for a page of your browser-based application to load

Reviewer 4: ```suggestion
  * How long does it take for a page of your browser-based application to load
  * How long it takes for an API system to respond to an API call
```

Author: Updated in Google doc.

  * How long does a database query take
  * How long does it take to provision some user-requested environment or
    resource

You will likely want to distinguish between the latency of successful and

> failed requests in your measurements and ultimatley create separate SLOs for

Reviewer 4: ```suggestion
failed requests in your measurements and ultimately create separate SLOs for
```

Author: Fixed in Google doc.

> both.

Reviewer 4: I disagree that there should be a latency SLO for failed requests. Failed requests should be handled by an availability SLO.

Reviewer: @davidffrench Time to fail is a fair thing to be concerned about (though I'm not sure it's a beginner's SLO). If a request is "doomed" (it's going to fail at some point in a request chain) then it makes a lot of sense to figure that out as soon as possible. Identifying "doomed" requests and notifying the requestor means the feedback cycle is shortened, and possibly less resource consumption on the service side.

In real terms: If a customer is asking for a cluster to be installed, but there's a fatal fault with their cloud account, we should have pre-flight checks in place so a potentially long install process doesn't abort after some long period of time where we keep the customer waiting.

Author: I copied this comment into the Google doc for further discussion.

### Errors

Errors measures the rate of failed requests to your service. Examples of how
to measure this might be:
  * The number of requests to your API that have a 5xx response code
  * The number of errors seen on the client-side by users of your web app
  * The rate of dropped packets on a network

### Saturation

Saturation measures the utilization of your service relative to that point at
which it would be considered "full". Effectively implementing SLIs for
saturation requires knowledge, or at least a hypothesis, of the capacity
for your service and the limiting factor. Examples of hor to measure this might
be:

  * CPU utilization percentage for services that are highly CPU-constrained
  * Memory utilization percentage for services that are highly Memory
    constrained
  * Storage capacity usage percentage
  * Network bandwidth capacity usage percentage
  * API request count as a percentage of maximum capacity

Work to identify at least one metric from each of these categories to
implement as SLIs for your service, then focus on identifying SLOs based on
them. Finally, expect to add to and refine your SLIs over time as you

> [mature](./sre_maturity.md) on your SRE journey.

Reviewer 4: Two points on this:

1. It would be better to move this above the descriptions of the four golden signals.
2. It might just be me but I find it much more useful to think about the SLOs first and then think of the SLIs that can back them. The SLOs should be thought of from the perspective of the end user and what the most important aspects of the level of service is to them. For example, in RHOSAK, we want to have a good first experience to the end users on Kafka creation so we defined a `correctness` SLO on how many Kafkas we expect to be created successfully and a creation `latency` SLO as the time it takes from user creation initiation to a ready Kafka instance is very important.

Author: I copied this comment into the Google doc for further discussion.
